File: array/python/GasStation.py
def canCompleteCircuit(gas, cost):
    if sum(gas) < sum(cost):    return -1
    n = len(gas)
    start = tank = 0
    for i in range(n):
        tank += gas[i]
        if tank < cost[i]:
            start = i + 1
            tank = 0
        else:   tank -= cost[i]
    return start

if __name__=='__main__':
    test_cases = [[[1,2,3,4,5], [3,4,5,1,2]],
                  [[2,3,4], [3,4,3]]]
    for gas, cost in test_cases:
        print(canCompleteCircuit(gas, cost))

File: greedy/java/src/GasStation.java
public class GasStation {

	public int canCompleteCircuit(int[] gas, int[] cost) {
		int sumGas = 0;
		for (int g : gas) {
			sumGas += g;
		}
		int sumCost = 0;
		for (int c : cost) {
			sumCost += c;
		}
		if (sumCost > sumGas) {
			return -1;
		}
		int start = 0;

		int currentGas = 0;
		for (int i=0; i<gas.length; i++) {
			currentGas += gas[i];
			if (currentGas < cost[i]) {
				start = start + 1;
				currentGas = 0;
			} else {
				currentGas -= cost[i];
			}
		}

		return start;
	}

	public static void main(String[] args) {
		int[] gas = new int[] {1, 2, 3, 4, 5};
		int[] cost = new int[] {3, 4, 5, 1, 2};
		GasStation solution = new GasStation();
		int result = solution.canCompleteCircuit(gas, cost);
		System.out.println(result);
	}
}
